### PR TITLE
BUG: Fix einsum optimize logic for singleton dimensions

### LINKED ISSUE #10343: einsum broadcast regression (with optimize=True)
In numpy 1.13.3, it was possible to execute the following snippet without errors, while in 1.14.0 this happens:
```python
In [1]: import numpy as np
In [2]: a = np.ones((10,2))
In [3]: b = np.ones((1,2))
In [4]: np.einsum('t...i,ti->t...', a, b)
Traceback (most recent call last):
  File "<ipython-input-4-fa62d1d882f9>", line 1, in <module>
    np.einsum('t...i,ti->t...', a, b)
  File "/usr/local/lib/python2.7/dist-packages/numpy/core/einsumfunc.py", line 1087, in einsum
    einsum_call=True)
  File "/usr/local/lib/python2.7/dist-packages/numpy/core/einsumfunc.py", line 710, in einsum_path
    "not match previous terms.", char, tnum)
ValueError: ("Size of label '%s' for operand %d does not match previous terms.", 't', 1)

```
However `optimize=False` solves the problem:
```python
In [5]: np.einsum('t...i,ti->t...', a, b, optimize=False)
Out[5]: array([2., 2., 2., 2., 2., 2., 2., 2., 2., 2.])
```
Is this intended behavior and the user is now responsible to explicitly disable optimization or is this a bug?

### PR DESCRIPTION
**Pull request by larsoner**
Closes #10343.

I don't know if it's the correct fix because these functions have a lot of logic going on that I have never looked at before (and haven't looked at too closely). Feel free to close if the right solution is a bit more involved.

### CODE REVIEW

**Comment by mhvk** [comment]
@larsoner - thanks for the PR. It took me a *long* time to see what the heck is going on in the parsing, but now I understand that, basically, the `optimize` path did not include broadcasting -- I must add it surprised me that broadcasting was even allowed on identical indices (rather than just on the `...`, but that seems to be the case before too).

Anyway, I think this PR is correct. @dgasmith - could you have a look?

**Comment by dgasmith** [comment]
I need to think on it a bit, that parsing tech is fairly complex as folks want quite the variety of input possibilities. I would consider adding more tests that cover the "double broadcasting" cases where 1-dims are mixed with ellipses like in the original issue.

**Comment by dgasmith** [comment]
This might hit issues in the BLAS calls. Please try the following two cases:

```python
a = np.random.rand(1, 5)
b = np.random.rand(5, 5)

np.einsum("...ij,...jk->...ik", a, a)
np.einsum("...ij,...jk->...ik", a, b)
```
I suspect we may have to go in and change the BLAS logic to bounce broadcasting cases as well.

**Comment by larsoner** [comment]
Rebased due to conflict and test added. I'm not on my dev machine now so I can't easily run the new tests so I'll let the CIs do the work :)

**Comment by larsoner** [comment]
@dgasmith @mhvk tests are indeed a bit broken now. It might take me a while to fix this, so feel free to take over, or I can find time in the coming weeks to work on it.

**Comment by dgasmith** [comment]
Yup, ill need to go in and really take a look at whats going on. Ill get to it in the next few days.

**Comment by larsoner** [comment]
Okay @dgasmith I deduplicated the code a bit, feel free to look to see if you're happy.

CIs are happy, ready for review/merge from my end.

**Review by dgasmith — APPROVED** [review]
Looks good to me, the NumPy maintainers might have a few convention thoughts.

**Comment by eric-wieser** [comment]
Can you loop me in on what the `MRG` prefix means? I've seen it used a bunch in matplotlib too

File: numpy/core/einsumfunc.py [hunk 1]
```diff
@@ -1101,13 +1104,22 @@ def einsum(*operands, **kwargs):
...
-
             # Checks have already been handled
             input_str, results_index = einsum_str.split('->')
             input_left, input_right = input_str.split(',')
-
+            if 1 in tmp_operands[0] or 1 in tmp_operands[1]:
+                left_dims = {dim: size for dim, size in
+                             zip(input_left, tmp_operands[0].shape)}
+                right_dims = {dim: size for dim, size in
+                              zip(input_right, tmp_operands[1].shape)}
```
Comment (eric-wieser) [inline numpy/core/einsumfunc.py]: Why not use `idx_rm` here? The other entries in this dict are being ignored

**Comment by larsoner** [comment]
> Can you loop me in on what the MRG prefix means? I've seen it used a bunch in matplotlib too

It's a way for me (contributor) to signal to you (reviewers) that the PR is ready for review / merge from my perspective. In other projects I work on, maintainers historically have then changed the title as they approve to e.g. `MRG+1`, `MRG+2` etc., though GitHub recently came out with the "approve" / "request changes" stuff so now that mechanism gets used more than additional title changes from the reviewer end.

File: numpy/core/einsumfunc.py (hunk 1 above)
Comment (larsoner) [reply numpy/core/einsumfunc.py]: Here `input_left` is e.g. `'ik'`, so this seems like a straightforward / quick way to get the indices to line up. Are you thinking something like this should be done instead?
```
if any(tmp_operands[0].shape[input_left.index(ind)] != tmp_operands[1].shape[input_right.index(ind)] for ind in idx_rm)
```

Comment (eric-wieser) [reply numpy/core/einsumfunc.py]: You're right, what you have is better.

File: numpy/core/einsumfunc.py [hunk 2]
```diff
@@ -1101,13 +1104,19 @@ def einsum(*operands, **kwargs):
...
+        # Handle broadcasting vs BLAS cases
         if blas:
-
             # Checks have already been handled
             input_str, results_index = einsum_str.split('->')
             input_left, input_right = input_str.split(',')
+            if 1 in tmp_operands[0] or 1 in tmp_operands[1]:
+                if any(tmp_operands[0].shape[input_left.index(ind)] !=
+                       tmp_operands[1].shape[input_right.index(ind)]
+                       for ind in idx_rm):
```
Comment (eric-wieser) [inline numpy/core/einsumfunc.py]: To reiterate my now-hidden comment, I think what you had before was better - sorry!

Comment (larsoner) [reply numpy/core/einsumfunc.py]: That's fine, easy enough to remove the commit

**Comment by charris** [comment]
Merged, thanks @larsoner .

**Pull request merged**